src/collection/tx_data/collector.py

```diff
@@ -0,0 +1,282 @@
+        except SolanaRpcException as e:
+            LOG.error(f"SolanaRpcException while fetching the last finalized block on chain: {e}")
+            time.sleep(0.5)
+            return self._get_latest_finalized_block_on_chain()
+
+    @log_performance_time(LOG)
+    def _write_tx_data(self) -> None:
+        """
+        Write raw tx data to database.
+        """
+        # if relevant transactions were not found in current iteration, do nothing.
+        if not self.rel_transactions:
+            self._report_collection()
+            return
+        with db.get_db_session() as session:
+            for transaction in self.rel_transactions:
+                sources = transaction.sources(self.protocol_public_keys if self.protocol_public_keys else [])
+                signature = transaction.signature
+                # Same transaction has to be recorded once per each relevant protocol
+                for source in sources:
+                    records = session.query(db.TransactionStatusWithSignature).filter_by(
+                        signature=str(signature),
+                        source=source
+                    ).all()
+
+                    # Check if the record exists.
+                    if records:
+                        # Iterate over each record and update transaction data.
+                        for record in records:
+                            record.transaction_data = transaction.tx_body.to_json()
+                    else:
+                        new_record = db.TransactionStatusWithSignature(
+                            source=source,
+                            slot=transaction.block_number,
+                            signature=str(signature),
+                            block_time=transaction.block_time,
+                            transaction_data=transaction.tx_body.to_json(),
+                            collection_stream=self.COLLECTION_STREAM
+                        )
+                        session.add(new_record)
+
+            # Commit the changes.
+            session.commit()
+        self._report_collection()
+
+    @abstractmethod
+    def _report_collection(self):
+        """
+        Report collected blocks.
+        :return:
+        """
+        raise NotImplementedError("Implement me!")
+
+    def _is_transaction_relevant(
+            self,
+            transaction: EncodedTransactionWithStatusMeta
+    ) -> bool:
+        """
+        Determines if the transaction involves any of the specified public keys.
+
+        This method checks if any of the public keys provided in the 'ppks' list
+        are present in the transaction's account keys. It is used to filter transactions
+        based on the involvement of specific protocols identified by their public keys.
+        """
+        for i in transaction.transaction.message.account_keys:  # type: ignore
+            if str(i.pubkey) in self.protocol_public_keys:  # type: ignore
```

Comment:
I think this can still return True for irrelevant transactions, but can't think of a workaround. Let's keep it like this.

Reply:
for sure. we dont know until we parse it